chaincfg+blockchain: abstract/refactor BIP 9 version bits implementation to work w/ BIP 8 block heights

Roasbeef opened the pull request:
In this PR, we further abstract our Version Bits implementation to be able to support _both_ BIP 9 time based timeouts, as well as BIP 8 block based timeouts. In order to bridge the two mechanisms, we introduces a series of new light interfaces that allow us to abstract away from when exactly a soft fork deployment "starts" and "ends". This PR doesn't yet add any new BIP 8 compatible deployments, as it's mean to be a pure refactoring that doesn't introduce any new behavior. It's also the case that BIP 8 has some extra Version Bits states that need to be added to our Version Bits implementation. 

The only thing missing from this PR is a few additional tests to confirm that no change in behavior has occurred. Reviewers should examine the change to how we define deployments that are "always started" and "never ended", as well as the usage of `time` methods to determine if a deployment has started or ended.

Roasbeef commented:
The alternative to this PR is to just go with the old: "lower than this value is a block height" approach which is used for lock time related logic elsewhere. However, I figured it's better to just further abstract things and hide more of the implementation details from the call sits within the `blockchain` package.

gchaincl commented on blockchain/versionbits.go:
```diff
@@ -123,27 +120,45 @@ type deploymentChecker struct {
+
+	// Can't fail as we make sure to set the clock above.
+	header := blkNode.Header()
+	started, _ := deploymentStarter.HasStarted(&header)
```
Since `deploymentStarter` is an interface, and you don't know how the the actual implementation looks like, wouldn't if be better to check for the error?

Roasbeef commented:
So for this to potentially be "taproot ready", assuming this "speed trial" thing happens, then we'll need to also add a `minimum_activation_height` parameter. Basically this would be a new gate here: https://github.com/btcsuite/btcd/blob/master/blockchain/thresholdstate.go#L243 to only go to active if that height has passed.

wpaulino commented on chaincfg/params.go:
```diff
@@ -69,13 +68,13 @@ type ConsensusDeployment struct {
+	// DeploymentStarter is used to determine if the given
+	// ConsensusDeployment has started or not.
+	DeploymentStarter ConsensusDeploymentStarter
 
-	// ExpireTime is the median block time after which the attempted
-	// deployment expires.
-	ExpireTime uint64
+	// DeploymentEnder is used to determine if the given
+	// ConsensusDeployment has ended or not.
+	DeploymentEnder ConsensusDeploymentEnder
```
This change also needs to be propagated to the RPC server (build errors below). We'll want to still expose `StartTime` and `EndTime` to populate the RPC response.

```
./rpcserver.go:1292:39: deploymentDetails.StartTime undefined (type chaincfg.ConsensusDeployment has no field or method StartTime)
./rpcserver.go:1293:39: deploymentDetails.ExpireTime undefined (type chaincfg.ConsensusDeployment has no field or method ExpireTime)
```

wpaulino reviewed (commented):
I really like this change, even though Taproot didn't activate through BIP 8, it paves the way forward for any future soft forks that do.

As you mentioned, this PR is still missing the state transition changes as part of Speedy Trial.

Roasbeef commented on chaincfg/deployment_time_frame.go:
```diff
@@ -0,0 +1,172 @@
+
+	// We check both after and equal here as after will fail for equivalent
+	// times, and we want to be inclusive.
+	return medianTime.After(m.startTime) || medianTime.Equal(m.startTime), nil
```
IIRC there was some edge case related to this we ran into in the past?

Roasbeef replied:
Nevertheless, tho this is more verbose, I prefer it as it's more explicit IMO (explicitly cover the == case).

Roasbeef commented on chaincfg/params.go:
```diff
@@ -365,19 +376,31 @@ var RegressionNetParams = Params{
-			ExpireTime: math.MaxInt64, // Never expires.
+			BitNumber: 1,
+			DeploymentStarter: NewMedianTimeDeploymentStarter(
+				time.Unix(0, 0), // Always available for vote
```
I think that was before i added the `IsZero` special case in the started/ended methods. Switched over to be more uniform.

Roasbeef commented:
Initial comments addressed, next commit with add the possibility for a custom threshold per deployment and also the min activation height. Will then add a new dummy deployment to test against that has both of these new values specified.

wpaulino commented on blockchain/versionbits.go:
```diff
@@ -123,27 +120,36 @@ type deploymentChecker struct {
+	// Can't fail as we make sure to set the clock above when we
+	// instantiate *BlockChain.
+	header := blkNode.Header()
+	started, _ := c.deployment.DeploymentStarter.HasStarted(&header)
```
I realize it can't fail as is right now, but we should handle the error anyway in case this is not caught/updated whenever a new fallible deployment implementation is added.

Roasbeef replied:
As is, starting can never fail. When/if a deployment that can fail here (which may come in a few years given the current rate of soft forks in the system), we can thread through the error. With the current function signature, not having to check the error makes the `thresholdStateTransition` more readable as there's less line noise.

Roasbeef commented:
Ok, this should be good to go now. Reviews should ensure they've internalized the commit body in this commit: https://github.com/btcsuite/btcd/pull/1700/commits/c1842e23c86b802d84edfc29d4c875982ce88f90

kcalvinalvin commented on ;:
```diff
@@ -0,0 +1,547 @@
+// Copyright (c) 2013-2017 The btcsuite developers
```
I think you added this file as a mistake.

Roasbeef replied:
Indeed...removed

Sjors commented:
From c1842e23c86b802d84edfc29d4c875982ce88f90:

> One thing worth noting here is that logic at the end of BIP 341 doesn't
match bitcoind as implemented. The bitcoind logic fixes an off-by-one
error in the BIP itself.

Just to make sure I understand what you're saying here. Are you referring to the transition from `LOCKED_IN` to `ACTIVE`?

BIP 341 [states](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#deployment):

```cpp
case LOCKED_IN:
        if (block.nHeight < min_activation_height) {
            return LOCKED_IN;
        }
        return ACTIVE;
```

The returned value (`LOCKED_IN` or `ACTIVE`) in this pseudo code refers to the state for the _current_ block, i.e. the block at `block.nHeight`. This follows the convention in BIP 9, where e.g. the pseudo code returns `DEFINED` for the genesis block.

So the current block is `ACTIVE` when its height is at or above the activation height.

Bitcoin Core does the [following](https://github.com/bitcoin/bitcoin/blob/v22.0/src/versionbits.cpp#L82-L85):

```cpp
case ThresholdState::LOCKED_IN: {
                // Progresses into ACTIVE provided activation height will have been reached.
                if (pindexPrev->nHeight + 1 >= min_activation_height) {
                    stateNext = ThresholdState::ACTIVE;
```

This is part of the function:
```cpp
    /** Returns the state for pindex A based on parent pindexPrev B. Applies any state transition if conditions are present.
     *  Caches state from first block of period. */
    ThresholdState GetStateFor(const CBlockIndex* pindexPrev,
```

When this function is called with the Tip block, it returns the state for the next (unmined!) block. So `pindexPrev->nHeight + 1` refers to the upcoming block, i.e. the rules that apply in the mempool. So the fork becomes `ACTIVE` for the mempool when `height of the tip block + 1` is `min_activation_height`.

So I don't think there's an off by one error in the BIP, but both the BIP and the Bitcoin Core implementation are confusing enough to be likely to cause off-by-one errors (and in fact they did with SegWit2x).

As an aside: afaik the deployment info was added to BIP 341 _after_ Taproot locked in with the speedy trial.

Roasbeef commented:
Decided ultimately to special case things a bit just so it matches 1:1, once this base PR is in, I want to follow up with some additional tests for the hybrid state machine.

Roasbeef commented:
Hey @Sjors, yeh I explained this on #btcd IRC. I was just initially tripped up by the fact that the BIP and the bitcoind implementation don't match up (one uses the current block (to be mined/accepted) while the other uses the previous. Our BIP 9 implementation always uses the previous block.

Roasbeef commented:
Implemented the BIP 9 modifications, and added a new unit test here: https://github.com/btcsuite/btcd/pull/1700/commits/75d3489f30dfe9282ea1505920dcd9090796f429. 

This PR should be g2g now.

guggero commented on chaincfg/deployment_time_frame.go:
```diff
@@ -0,0 +1,181 @@
+// started. A deployment has started once according to the current "time", the
+// deployment is eligible for activation once a perquisite condition has
+// passed.
+type ConsensusDeploymentStarter interface {
```
Is there ever a deployment that has a different mechanism/implementation for starting and ending? Or could this just be a `ConsensusActivation` with two methods and the extended `ClockConsensusActivation` interface and `MedianTimeActivation` implementation... Would make code in later commits a bit simpler. But non-blocking.

Roasbeef replied:
> Is there ever a deployment that has a different mechanism/implementation for starting and ending? 

Back when this was made, doing taproot activation with a combination of height/time was an open option. Since then time was used for start/end and height for the min time. The refactors here now aren't _totally_ necessary, but certainly made some of the added unit tests I wrote a bit easier.

guggero commented on chaincfg/params.go:
```diff
@@ -94,6 +94,19 @@ type ConsensusDeployment struct {
+	// existing RuleChangeActivationThreshold value set at the
+	// network/chain level. This value divided by the active
+	// MinerConfirmationWindow denotes the threshold required for
+	// activation. A value of 1850 block denotes a 90% threshold.
```
Should this say `1815` instead?

guggero commented on chaincfg/params.go:
```diff
@@ -415,6 +440,17 @@ var RegressionNetParams = Params{
+		DeploymentTestDummyMinActivation: {
+			BitNumber:                 22,
+			CustomActivationThreshold: 72,  // Only needs 50% hash rate.
+			MinActivationHeight:       600, // Can only activate after height 600.
```
Not sure I understand this correctly, but does this mean Taproot would only activate after 600 blocks on regtest? That would require us to mine quite a few blocks during integration tests and similar tasks.

Roasbeef replied:
For tapproot, we'd set a height of zero so it can activate at anytime. This new deployment was added so I could test the "speedy trial" specific additions like the min activation height. 

This PR removed the old taproot deployment, as it can actually cause a panic in certain cases, as it needs to be defined for all network types (was only defined for segnet before).

guggero commented on blockchain/thresholdstate.go:
```diff
@@ -248,17 +264,37 @@ func (b *BlockChain) thresholdState(prevNode *blockNode, checker thresholdCondit
+			// stay in the locked in position.
+			if !checker.EligibleToActivate(prevNode) {
+				state = ThresholdLockedIn
+
```
nit: extra newline. Or add `break` to continue without indentation below?

guggero commented on integration/bip0009_test.go:
```diff
@@ -268,7 +272,33 @@ func testBIP0009(t *testing.T, forkKey string, deploymentID uint32) {
+	// over.
+	numBlocksLeft := confirmationWindow
+	for i := uint32(0); i < numBlocksLeft; i++ {
+		_, err := r.GenerateAndSubmitBlock(nil, signalForkVersion,
```
nit: formatting

guggero commented on integration/bip0009_test.go:
```diff
@@ -268,7 +272,33 @@ func testBIP0009(t *testing.T, forkKey string, deploymentID uint32) {
+			time.Time{})
+		if err != nil {
+			t.Fatalf("failed to generated block %d: %v", i, err)
+		}
```
Before generating and submitting a block, should we assert that we're still in state `blockchain.ThresholdLockedIn`? This should be true until after we mine the very last block (after which we exit the loop).

Roasbeef replied:
Good idea, added on.

guggero reviewed (commented):
I have limited understanding of activation mechanisms in general, but from what I see in this diff the changes make sense to me. Always nice to see how a well-defined interface can make things very straightforward to understand!

guggero reviewed (approved):
LGTM :tada:

wpaulino commented:
Post-merge ACK 👍
